Adding action functions for NPCs (KIND 5, KIND 6, KIND 7)

junasano opened the pull request:
Actions is an aspect of the NPC module which allows the player to interact with the NPCs by talking to and trading with them. The current actions can be split into three larger groups, called KIND 5, KIND 6, and KIND 7 actions. 

The purpose of this PR is to implement three functions which allow these KIND actions to be processed. Also, a new variable called price was added to the item_t struct (which is necessary for the KIND 7 function)

KIND 5 and KIND 6 are fully complete. However, with KIND 7, the BUY part cannot be implemented yet since there needs to be a currency-like feature first. Additionally, for TRADE for KIND 7 the player can’t choose which item to give away and it just takes the first item on the list which has a higher price than the wanted item. These two issues will be documented and issues will be created on the backlog later. 
Closes #1574

Commit:
rough implementation complete

Commit:
merged dev branch

Commit:
Merge branch 'dev' of github.com:uchicago-cs/chiventure into npc/actions_final

Commit:
Merge branch 'dev' of github.com:uchicago-cs/chiventure into npc/actions_final

Commit:
made changes for previous PR on this one by accident

Commit:
merged from dev branch

Commit:
fixed small bug

Commit:
added contains_action check for kind 6 and 7

Commit:
fixed minor bug

Commit:
added price variable to item struct to use for trading

srydbergcox commented on src/action_management/src/actionmanagement.c:
```diff
@@ -411,3 +413,305 @@ int do_self_action(chiventure_ctx_t *c, action_type_t *a,
+                return CONDITIONS_NOT_MET;
+            }
+            // initiates conversation (set_game_mode to CONVERSATION)
+            int switch_mode; // check in examples if this is how a convo is initiated
```
Was this checked? I am not familiar with this mechanism so want to be sure.

sglenn-uchi replied:
Hmm I checked but might be worth looking at again. Jun have you by chance looked at it?

junasano replied:
Just checked again. And this was correct so I removed the comment. 
But when the mode is switched to NORMAL, the third parameter was NULL for the examples so I changed it to NULL on line 525

sglenn-uchi replied:
Thank you for catching that!

srydbergcox commented on src/action_management/src/actionmanagement.c:
```diff
@@ -411,3 +413,305 @@ int do_self_action(chiventure_ctx_t *c, action_type_t *a,
+        {   
+            return FAILURE;
+        }
+        if(add_item_to_npc(agent->npc, item) != SUCCESS)
```
Does this function call and the function call on line 588 remove and add the item too? Or, do we need to directly call the functions too in order to update the inventories?

sglenn-uchi replied:
I don't think any direct calls are necessary– add_item_to_npc adds the given item to the hash table representing the npc's inventory and then just returns whether it was able to complete successfully or not. So I don't think direct calls are necessary, if that's what you're talking about.

srydbergcox replied:
Yes, that is what I was wondering about. Sounds good then, I just wanted to be sure since I wasn't sure how the functions worked under-the-hood.

srydbergcox commented on src/action_management/src/actionmanagement.c:
```diff
@@ -411,3 +413,305 @@ int do_self_action(chiventure_ctx_t *c, action_type_t *a,
+        {
+            return FAILURE;
+        }
+        if(add_item_to_player(c->game->curr_player, item, c->game) != SUCCESS)
```
Same question as above for lines 612 & lines 608.

srydbergcox reviewed (changes requested):
These changes look good and I think it is a really great addition to Chiventure! My only comments are in the main function that you wrote where you call functions that add items, etc. in a conditional check statement and I just want to be sure that that actually updates the player's inventories. Just for now, I am going to hold off on approving the PR until we can discuss that real quick since if they didn't update the inventories the functions wouldn't work. Other than that though, great work!

srydbergcox reviewed (approved):
Looks good!

Commit:
Merge branch 'dev' of github.com:uchicago-cs/chiventure into npc/actions_final

Commit:
removed comment and fixed small mistake

sglenn-uchi committed:
Merge branch 'dev' into npc/actions_final

Commit:
merged dev branch

Don-Osipov commented on include/action_management/action_structs.h:
```diff
@@ -35,16 +35,49 @@ enum actions {
+
+/* Each enum corresponds to a different object of the self action type.
+   e.g. The STATS in VIEW STATS*/
+typedef enum {
```
As someone from the action management team I am a bit confused on what this enum does and how it is used. Perhaps update the documentation to make it more clear or give a more concrete example?

junasano replied:
Thanks for pointing this out. I actually don't know what this is and deleting it doesn't seem to cause any issues so I have removed it. I'm assuming I accidentally copied this from the old branch when I was moving stuff around.

Don-Osipov commented on src/action_management/src/actionmanagement.c:
```diff
@@ -411,3 +413,304 @@ int do_self_action(chiventure_ctx_t *c, action_type_t *a,
+                    {   
+                        return FAILURE;
+                    }
+                    if(add_item_to_player(c->game->curr_player, item, c->game) != SUCCESS)
```
the "c->game->curr_player" object is used many times in the functions you wrote. I think assigning it to a variable would make the code more readable. In general, just try to utilize named variables more often across your functions.

Don-Osipov reviewed (commented):
Overall the code you guys wrote looks great. It is well documented and structured. I just had some minor nitpicks with documentation for self_action_object and also the repeated use of c->game->curr_player without assigning it to a variable.

Commit:
Merge branch 'dev' of github.com:uchicago-cs/chiventure into npc/actions_final

Commit:
styling issue and fixed typo on test_class_prefabs

Commit:
creating a c_player variable to make it more readable

junasano commented:
> Overall the code you guys wrote looks great. It is well documented and structured. I just had some minor nitpicks with documentation for self_action_object and also the repeated use of c->game->curr_player without assigning it to a variable.

Thanks for the review! I deleted the sefl_action_object and implemented your second suggestion.

Don-Osipov reviewed (approved):
All feedback that I gave before was addressed. Nice work!

Commit:
Merge branch 'dev' of github.com:uchicago-cs/chiventure into npc/actions_final

lilyehsani reviewed (approved):
LGTM! Thanks to the other thorough reviews I have no notes.

lilyehsani committed:
Merge branch 'dev' into npc/actions_final

lilyehsani committed:
Merge branch 'dev' into npc/actions_final

lilyehsani commented:
Issue Score: Excellent

Comments:
Great work! Like issues, PRs should have closing statements.

<!-- {"score": "excellent", "minor": false} -->
